core/templates/dev/head/domain/topic/TopicObjectFactory.js

```diff
@@ -112,7 +121,7 @@ oppia.factory('TopicObjectFactory', ['SubtopicObjectFactory',
         if (isNewlyCreated && subtopicDeleted) {
           this._subtopics[i].decrementId();
         }
-        if (this._subtopics[i].getId() === subtopicId) {
+        if ((this._subtopics[i].getId() === subtopicId) && (!subtopicDeleted)) {
```

Comment:
Hm ... why would the subtopic be deleted before you get to it? Not sure I understand why the second clause was added.

Reply:
Since I am decrementing the subtopicId above, the new decremented ids of subsequent subtopics can become equal to the id of the deleted subtopic

Reply:
Oh I see. That sounds a bit risky, actually. If a developer needs to think hard to follow the logic, then the function is too complicated.

How about: in a first pass, find the subtopic and delete it, making a note of its ID. Then do a second pass to update the IDs of any subtopics whose IDs come after the deleted one. Would that work? (Acknowledged that it's half as efficient, but the number of subtopics is typically small so I don't see this being an issue.)

Reply:
Ok, yeah, that would be fine.

Reply:
Done.